src/UI/Components/Diagram/helpers/relations.ts

```diff
@@ -1,47 +1,46 @@
 import { dia } from "@inmanta/rappid";
-import {
-  EmbeddedEntity,
-  InstanceAttributeModel,
-  InterServiceRelation,
-  ServiceInstanceModel,
-  ServiceModel,
-} from "@/Core";
+import { EmbeddedEntity, InterServiceRelation, ServiceInstanceModel, ServiceModel } from "@/Core";
 import { ServiceEntityBlock } from "../shapes";
 
 /**
  * Extracts the IDs of the relations of a service instance.
  *
- * @param service - The service model.
- * @param instance - The service instance.
- * @returns {string[]} An array of relation IDs.
+ * Prioritizes candidate_attributes over active_attributes when both are available.
+ * Filters out null, undefined, and empty string values.
+ *
+ * @param service - The service model containing inter-service relation definitions.
+ * @param instance - The service instance containing attribute values.
+ * @returns {string[]} An array of relation IDs as strings.
  */
 export const extractRelationsIds = (
   service: ServiceModel,
   instance: ServiceInstanceModel
 ): string[] => {
   const relationKeys = service.inter_service_relations.map((relation) => relation.name);
 
-  if (!relationKeys) {
+  if (relationKeys.length === 0) {
     return [];
   }
 
-  const extractRelation = (attributes: InstanceAttributeModel): string[] =>
-    relationKeys
-      .map((key) => String(attributes[key]))
-      .filter((attribute) => attribute !== "undefined");
+  // Prefer candidate_attributes over active_attributes.
+  // rollback_attributes are not used in the composer.
+  const attributes = instance.candidate_attributes ?? instance.active_attributes;
 
-  if (instance.candidate_attributes !== null) {
-    return extractRelation(instance.candidate_attributes);
-  } else if (instance.active_attributes !== null) {
-    return extractRelation(instance.active_attributes);
-  } else {
+  if (!attributes) {
     return [];
   }
+
+  return relationKeys
+    .map((key) => attributes[key])
+    .filter(Boolean)
+    .map(String);
```

Comment:
what does this actually do? Specifically the filter and the map

Reply:
I could be wrong, but if this behaves the same as it does in python, this takes a list of keys (`relationKeys`) as input and for each of these key, we look at the associated value in a map (`attributes[key]`), if it is truthy, we cast it to string and add it to the output list, if it is falsy, we drop it.

- filter will apply a function on each element and only keep those that pass the test
- map will apply a given function on each element and replace them with the result

this looks like a bug compared to previous implementation. I think this now only keeps the truthy attributes (and drops all falsy ones e.g. "false" ) whereas before we would drop only the undefined ones.

Reply:
Correct. 
But the bug was the previous implementation, where empty values, or null values would be retained, which wouldn't be valid ids.
But looking for implementations, I realize the method is not used anymore 😅 Should have started with that before making a change to it...